### PR TITLE
Mfa auth operation

### PR DESCRIPTION
**Pull request by mary-dcouto**

### CODE REVIEW

File: Samples/Service/ZitiInstance.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenZiti;
+
+namespace OpenZiti
+{
+
+	public struct ZitiInstance
```
Comment (dovholuknf) [inline Samples/Service/ZitiInstance.cs]: This feels like it's not serving any purpose? I see it used in the Weather sample for what seems like MFA stuff... Let's move this into a "MFA sample" of its own

File: Samples/Service/ZitiTunnelCommand.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OpenZiti
+{
+	public class ZitiTunnelCommand
```
Comment (dovholuknf) [inline Samples/Service/ZitiTunnelCommand.cs]: This is the "csharp sdk" not a "tunneling sdk"... I don't think words like "tunnel" fit in this project. Please find/replace tunnel with something that's more accurate

File: Samples/Service/ZitiTunnelCommand.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	{
+		public struct NextAction
+		{
```
Comment (dovholuknf) [inline Samples/Service/ZitiTunnelCommand.cs]: on the whole, find/replace tabs with spaces to follow the existing convention

File: Samples/Weather.cs [hunk 4]
```diff
@@ -19,28 +19,118 @@
 using System.Linq;
 using System.Text;
 using System.Text.Json;
+using System.Runtime.InteropServices;
 
 namespace OpenZiti.Samples {
+
     public class Weather {
         static MemoryStream ms = new MemoryStream(2 << 16); //a big bucket to hold bytes to display contiguously at the end of the program
+        static ZitiTunnelCommand.Options tunOptions = new ZitiTunnelCommand.Options();
```
Comment (dovholuknf) [inline Samples/Weather.cs]: extract to an MFA sample and i'll review again. keep the existing samples intact

File: Samples/Weather.cs [hunk 5]
```diff
@@ -19,28 +19,118 @@
...
+                    }
+                case 3:
+                    {
+                        Console.WriteLine("Remove MFA for the identity" + (zitiInstance.Zid?.IdentityNameFromController != null ? zitiInstance.Zid?.IdentityNameFromController : zitiInstance.Zid?.InitOpts.IdentityFile));
+                        Console.WriteLine("Enter the mfa auth codo: ");
+                        mfacode = Console.ReadLine();
+                        ZitiIdentity.TunnelCB tunnelCB = new ZitiIdentity.TunnelCB();
+                        tunnelCB.zidOpts = zitiInstance.Zid.InitOpts;
+                        ZitiIdentity.TunnelCB.ZitiResponseDelegate cbDelegate = tunnelCB.ZitiResponse;
+                        ZitiMFAService.remove_mfa(zitiInstance.Zid, mfacode, Marshal.GetFunctionPointerForDelegate(cbDelegate));
```
Comment (dovholuknf) [inline Samples/Weather.cs]: i'll make the same comments below probably but "ZitiMFAService" is a bad name and probably all these functions just belong on API.cs

File: Samples/Weather.cs [hunk 6]
```diff
@@ -19,28 +19,118 @@
...
+			}
+		}
 
         public static void Run(string identityFile) {
+            tunOptions.OnNextAction += OnZitiTunnelNextAction;
+            zitiInstance.Initialize();
+
             ZitiIdentity.InitOptions opts = new ZitiIdentity.InitOptions() {
-                EventFlags = ZitiEventFlags.ZitiContextEvent | ZitiEventFlags.ZitiServiceEvent,
+                EventFlags = ZitiEventFlags.ZitiContextEvent | ZitiEventFlags.ZitiServiceEvent | ZitiEventFlags.ZitiMfaAuthEvent | ZitiEventFlags.ZitiAPIEvent,
```
Comment (dovholuknf) [inline Samples/Weather.cs]: Can we make a `ZitiEventFlags.All`?

Comment (dovholuknf) [reply Samples/Weather.cs]: also - undo this particular commit to keep the samples 'as is' (and the ones below probably too, of course)

File: TestProject/TestProject.csproj [hunk 7]
```diff
@@ -18,15 +18,18 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x86'">
     <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
+
   <Target Name="x86" BeforeTargets="Build">
-      <Message Importance="High" Text="COPYING $(SolutionDir)\ZitiNativeApiForDotnetCore\build-win\$(Platform)\library\$(Configuration)\ziti4dotnet.dll to $(OutDir)" />
-      <Copy SourceFiles="$(SolutionDir)ZitiNativeApiForDotnetCore\build-win\$(Platform)\library\$(Configuration)\ziti4dotnet.dll" DestinationFolder="$(OutDir)" />
-      <Copy SourceFiles="$(SolutionDir)ZitiNativeApiForDotnetCore\build-win\$(Platform)\library\$(Configuration)\ziti4dotnet.pdb" DestinationFolder="$(OutDir)" />
+      <Message Importance="High" Text="COPYING $(SolutionDir)\..\ZitiNativeApiForDotnetCore\build-win\$(Platform)\library\$(Configuration)\ziti4dotnet.dll to $(OutDir)" />
```
Comment (dovholuknf) [inline TestProject/TestProject.csproj]: let's not escape from the solutiondir. probably should rewind this change

File: Ziti.NET.Standard/src/OpenZiti/API.cs [hunk 8]
```diff
@@ -239,4 +239,13 @@ public struct posture_query {
     public struct ziti_process {
         public string path;
     }
+
+    public struct ziti_identity
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/API.cs]: probably can remove this? i don't think it's necessary?

File: Ziti.NET.Standard/src/OpenZiti/ExtensionMethods.cs [hunk 9]
```diff
@@ -38,6 +38,10 @@ public static string GetDescription(this Enum enumVal)
         {
             var type = enumVal.GetType();
             var memberInfo = type.GetMember(enumVal.ToString());
+            if (memberInfo == null || memberInfo.Length == 0)
+			{
+                return NO_DESC;
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ExtensionMethods.cs]: I don't think we should plaster over this and I'd like to unwind this. if the description isn't returned we need to know what description and why. i don't think we should commit this change

File: Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs [hunk 10]
```diff
@@ -178,7 +178,8 @@ public class ZitiIdentity {
 			InitOpts.OnZitiContextEvent += SaveNativeContext;
 			StructWrapper ziti_opts_ptr = new StructWrapper(ziti_opts);
 
-			Native.API.ziti_init_opts(ziti_opts_ptr.Ptr, Loop.nativeUvLoop);
+			int result = Native.API.ziti_init_opts(ziti_opts_ptr.Ptr, Loop.nativeUvLoop);
+			Logger.Info("Result of ziti initialization : " + result);
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs]: remove the logging. instead test the result and throw a runtime exception (pick a type that makes sense) if the init fails

File: Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs [hunk 11]
```diff
@@ -259,14 +270,35 @@ public class ZitiIdentity {
...
+					Logger.Info("Ziti identifier received API event with controller address {0}", ziti_api_event.new_ctrl_address);
 					break;
 				default:
 					Logger.Warn("UNEXPECTED ZitiEventFlags [{0}]! Please report.", type);
 					break;
 			}
 		}
 
-		private void SaveNativeContext(object sender, ZitiContextEvent e) {
+		internal void SaveNativeContext(object sender, ZitiContextEvent e) {
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs]: needed as internal? revert to private?

File: Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs [hunk 12]
```diff
@@ -295,10 +330,38 @@ public struct InitOptions {
...
+			public void ZitiResponse(object evt)
+			{
+				if (evt is ZitiMFAStatusEvent)
+				{
+					zidOpts.ZitiMFAStatusEvent((ZitiMFAStatusEvent)evt);
+				}
+			}
+		}
 
+		public void native_ziti_pq_mac_cb(IntPtr ziti_context, string id, Native.ziti_pr_mac_cb response_cb) {
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs:394]: `native_ziti_pq_mac_cb` doesn't belong on this class. if this is necessary it needs to be converted to idiomatic csharp.

File: Ziti.NET.Standard/src/OpenZiti/ZitiMFAService.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	public struct ZitiMFAEnrollment
+	{
+		public bool isVerified;
+		public string[] recoveryCodes;
+		public string provisioningUrl;
+	}
+
+	public class ZitiMFAService
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ZitiMFAService.cs]: refactor and i'll review after

File: Ziti.NET.Standard/src/OpenZiti/API.cs (hunk 8 above)
Comment (mary-dcouto) [reply Ziti.NET.Standard/src/OpenZiti/API.cs]: I am using this struct to fetch the name of the identity inside the event_cb function 
ziti_identity zid = Marshal.PtrToStructure<ziti_identity>(idPtr)

File: Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs (hunk 12 above)
Comment (mary-dcouto) [reply Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs:394]: this function was already there in this cs file. I just added a logger to see if this gets invoked. I will fix this function when I work on the posture queries

File: Ziti.NET.Standard/src/OpenZiti/ZitiMFAService.cs (hunk 13 above)
Comment (mary-dcouto) [reply Ziti.NET.Standard/src/OpenZiti/ZitiMFAService.cs]: I think I will move the MFA operations to the ZitiIdentity class, so the users can call the MFA operations using the ZitiIdentity object. Is that okay?

File: Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs (hunk 12 above)
Comment (dovholuknf) [reply Ziti.NET.Standard/src/OpenZiti/ZitiIdentity.cs:394]: sure. it was most likely like committed as an afterthought by me. something i stubbed out to be 'fixed' later. we should either remove it entirely or 'fix' it is all i am trying to say. looking at the struct there are 3 other callbacks that would need the same kind of treatment. they seem to go unused though? Or maybe i just don't see the other callbacks used?

File: Ziti.NET.Standard/src/OpenZiti/ZitiStatus.cs [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright NetFoundry Inc.
+Copyright 2019-2020 NetFoundry, Inc.
```
Comment (dovholuknf) [inline Ziti.NET.Standard/src/OpenZiti/ZitiStatus.cs]: this is the wrong copyright. please revert to "Copyright NetFoundry Inc.". thx

Comment (mary-dcouto) [reply Ziti.NET.Standard/src/OpenZiti/ZitiStatus.cs]: done

**Review by dovholuknf — APPROVED** [review]

**Pull request merged**